### ISSUE TITLE
discrete/integer distributions cause warnings/errors

### ISSUE THREAD

**Issue by paciorek** [nimble-dev/nimble#705]
The dZIP example on our examples page used to work smoothly.  Now nimbleModel issues some errors or warnings -- not clear if it had actually broken or was just yelling.  Then it does appear to work, but not sure if getParam would work--the source of one of the error/warning messages.  

Also look into creating a test case for a discrete/integer user-defined distribution.

The example code is in the nimble-demos repo.

**Comment by paciorek** [comment]
Also, the examples page on website is out of date in that we don't need rZIP and don't need explicit registration of user-defined fxns.

**Comment by paciorek** [comment]
Ok, so this is not an issue with our checking. It is a more fundamental issue with how getParam() (and getBound()) interact with registerDistributions.

If a user provides a distribution that uses integers then registerDistributions sets the relevant 'type' to be integer. But we only create double versions of getParam.

Rheww options:
1) ignore the fact that there are integers in the d and r functions and set the type in distributions list via registerDistributions to always be double.
2) keep the 'correct' type  in the distributions list but in getParam and makeParamInfo replace 'integer' with 'double'.
3) create integer versions of getParam. I haven't looked yet to see why this is not happening already. But I think at some point we assume that only double versions are needed. 

I think I like option 2 best as it preserves the type information and then accounts for the fact that we only create double versions of getParam. Regarding option 3, given we always work with doubles, seems to make sense to continue doing that.

Side note (question for @perrydv ):  if a user creates a integer-valued distribution, we get RC functions for the d and r functions that use integers. But the calculate nodeFunction passes in doubles.
Somehow (the further we go in this project, the more I realize how little C++ I actually know...) this is not causing a compilation error. Here is an example:

```
dbernn <- nimbleFunction(
    run = function(x = integer(0), prob = double(0), 
        log = integer(0, default = 0)) {
        returnType(double(0))
        logProb <- x*log(prob) + (1-x)*log(1-prob)
        if(log) return(logProb)
        else return(exp(logProb)) 
    })

rbernn <- nimbleFunction(
    run = function(n = integer(0), prob = double(0)) {
        returnType(integer(0))
        if(n != 1) print("rmyexp only allows n = 1; using n = 1.")
        return(rbinom(1, 1, prob))
    })

code <- nimbleCode({
    y ~ dbernn(p)
})

model = nimbleModel(code, inits = c(y=0,p=0.5)) # ,check=F,calculate=F)
cm = compileNimble(model)
```
The resulting C++ code has this:
```
// note the int first arg
double  rcFun_1 ( int ARG1_x_, double ARG2_prob_, int ARG3_log_ )  {
double logProb;
...
}

double Interm_2;
double Interm_3;
Interm_2 = (**model_y)[0];
Interm_3 = (**model_p)[0];
// note use of a double as first arg
(**model_logProb_y)[0] = rcFun_1(Interm_2, Interm_3, 1);
```

Why does this compile and run without explicit type casting?

**Comment by perrydv** [comment]
@paciorek Thanks for laying this all clearly.  I agree option 2 makes sense.  If we pull off the redesign one day, I think it should be better with different types in models, but for now we do only ever implement numbers as doubles.

C++ will automatically cast if it knows how to.  For casting from double to integer, it does know how.  When one writes a class, one can define such casting operations.

**Comment by paciorek** [comment]
Fixed by PR #714
